drivers/eeprom/eeprom_st25dv.c

```diff
@@ -0,0 +1,970 @@
+	if (ret == 0) {
+		ret = eeprom_st25dv_write_end_zone(dev, EEPROM_ST25DV_END_ZONE2, end_value);
+	}
+
+	if (ret == 0) {
+		ret = eeprom_st25dv_write_end_zone(dev, EEPROM_ST25DV_END_ZONE1, end_value);
+	}
+
+	return ret;
+}
+
+int eeprom_st25dv_create_user_zone(const struct device *dev, const uint16_t zone1_length,
+				   const uint16_t zone2_length, const uint16_t zone3_length,
+				   const uint16_t zone4_length)
+{
+	uint8_t end_value;
+	uint16_t zone1_length_tmp = zone1_length;
+	uint16_t zone2_length_tmp = zone2_length;
+	uint16_t zone3_length_tmp = zone3_length;
+	uint16_t max_mem_length;
+	int ret = 0;
+	eeprom_st25dv_mem_size_t memsize;
+
+	memsize.mem_size = 0;
+	memsize.block_size = 0;
+
+	/* Get EEPROM mem size */
+	(void)eeprom_st25dv_read_mem_size(dev, &memsize);
+	max_mem_length = ((uint32_t)memsize.mem_size + 1u) * ((uint32_t)memsize.block_size + 1u);
+
+	/* Checks that values of different zones are in bounds */
+	if ((zone1_length < 32u) || (zone1_length > max_mem_length) ||
+	    (zone2_length > (max_mem_length - 32u)) || (zone3_length > (max_mem_length - 64u)) ||
+	    (zone4_length > (max_mem_length - 96u))) {
+		ret = -EINVAL;
+	}
+
+	/* Checks that the total is less than the authorised maximum */
+	if ((zone1_length + zone2_length + zone3_length + zone4_length) > max_mem_length) {
+		ret = -EINVAL;
+	}
+
+	/* if the value for each Length is not a multiple of 32 correct it. */
+	if ((zone1_length % 32u) != 0u) {
+		zone1_length_tmp = zone1_length - (zone1_length % 32u);
+	}
+
+	if ((zone2_length % 32u) != 0u) {
+		zone2_length_tmp = zone2_length - (zone2_length % 32u);
+	}
+
+	if ((zone3_length % 32u) != 0u) {
+		zone3_length_tmp = zone3_length - (zone3_length % 32u);
+	}
+
+	if (ret == 0) {
+		/* First right 0xFF in each Endx value */
+		ret = eeprom_st25dv_init_end_zone(dev);
+	}
+
+	if (ret == 0) {
+		/* Then Write corresponding value for each zone */
+		end_value = (uint8_t)((zone1_length_tmp / 32u) - 1u);
+		ret = eeprom_st25dv_write_end_zone(dev, EEPROM_ST25DV_END_ZONE1, end_value);
+	}
+
+	if (ret == 0) {
+		end_value = (uint8_t)(((zone1_length_tmp + zone2_length_tmp) / 32u) - 1u);
+		ret = eeprom_st25dv_write_end_zone(dev, EEPROM_ST25DV_END_ZONE2, end_value);
+	}
+
+	if (ret == 0) {
+		end_value =
+			(uint8_t)(((zone1_length_tmp + zone2_length_tmp + zone3_length_tmp) / 32u) -
+				  1u);
+		ret = eeprom_st25dv_write_end_zone(dev, EEPROM_ST25DV_END_ZONE3, end_value);
+	}
+
+	return ret;
+}
+
+static const struct eeprom_driver_api eeprom_st25dv_api = {
+	.read = eeprom_st25dv_read,
+	.write = eeprom_st25dv_write,
+	.size = eeprom_st25dv_size,
+};
+
+#define ST25DV_DEVICE_INIT(inst)                                                                   \
+	static const struct eeprom_st25dv_config _st25dv_config_##inst = {                         \
+		.bus = I2C_DT_SPEC_INST_GET(inst),                                                 \
+		.addr = DT_INST_REG_ADDR(inst),                                                    \
```

Comment:
What is the use of `.addr`?

Reply:
.addr is used as I2C address when calling i2c driver functions. Sometimes, it is used with | 4 because some command require bit 2 to be set for configuration purposes.